rust/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+
+pub use ntp::*;
+pub mod ntp;
+
+pub use tls::*;
+pub mod tls;
+
+use tls_parser::tls_ciphers::CIPHERS;
+
+type ProbeFn      = extern "C" fn (*const i8, u32, *const i8) -> u16;
+type ParseFn      = extern "C" fn (u8, *const i8, u32, *mut c_void) -> u32;
+type NewStateFn   = extern "C" fn () -> *mut c_void;
+type FreeStateFn  = extern "C" fn (*mut c_void) -> ();
+
+struct RustParser {
+    /// Protocol name.
+    name:         CString,
+    /// Default port
+    default_port: CString,
+
+    /// The structure exposed to the C code
+    c_parser:     Option<RustCParser>,
+}
+
+#[repr(C)]
+struct RustCParser {
+    /// Protocol name. Must be \0-terminated
+    name:         *const c_char,
+    ip_proto:     u16,
+    default_port: *const c_char,
+    min_frame_length: i32,
+    /// Application layer protocol ID
+    al_proto:     u16,
+    /// Events table
+    events:       *const c_void,
+    probe:        ProbeFn,
+    parse:        ParseFn,
+    new_state:    NewStateFn,
+    free_state:   FreeStateFn,
+}
+
+/// Declare RustCParser as shareable between threads.
+/// This is only true because we only use read-only instances, but is necessary to initialize
+/// the global registry and send structures to the C code.
+unsafe impl Sync for RustCParser { }
+
+extern {
+    fn crust_register_alproto(alproto: *const c_char) -> u16;
+}
+
+impl RustParser {
+    pub fn new(name: &str, proto: u16, default_port: &str, min_frame_length: i32,
+               global_al_proto: &mut u16,
+               events: *const c_void,
+               probe: ProbeFn, parse: ParseFn, new: NewStateFn, free: FreeStateFn)
+            -> RustParser {
+        let r = RustParser{
+            name:         CString::new(name).unwrap(),
+            default_port: CString::new(default_port).unwrap(),
+            c_parser:     None,
+        };
+        let al_proto = unsafe{ crust_register_alproto(r.name.as_ptr()) };
+        if (al_proto == 0) || (al_proto == 0xffff) {
+            panic!("Application layer protocol registration failed for protocol {}", name);
+        };
+        // println!("Registered alproto {} returned {}", name, al_proto);
+        *global_al_proto = al_proto;
+        RustParser{
+            c_parser: Some(
+                          RustCParser{
+                              name:         r.name.as_ptr(),
+                              ip_proto:     proto,
+                              default_port: r.default_port.as_ptr(),
+                              min_frame_length: min_frame_length,
+                              al_proto:     al_proto,
+                              events:       events,
+                              probe:        probe,
+                              parse:        parse,
+                              new_state:    new,
+                              free_state:   free,
+                          }
+                          ),
+            .. r
+        }
+    }
+}
+
+lazy_static! {
```

Comment:
It allows to define a static (immutable) structure (the hash table), using initialization code. If you don't use lazy_static, you can't use code, only declaration of values